### PR TITLE
Ch/challenge length

### PR DESCRIPTION
**Pull request by chalkandpaste**
# Goals of PR

Rewrite the `Challenge.sol` contract to have defender and challenger propose length of challenge and the final hash for the challenge, so that the pair which has the least length is chosen and bisection happens from there.

Core changes:

- Remove the `initializeChallengeLength` function in favor of `proposeChallenge`. `proposeChallenge` now takes proposals for the values of`numSteps` and `endStateHash` from both parties and chooses one such that the pair has the least `numSteps` of the two.
- Cleanup changes and add a few comments for readability.

Notes:

- Need to implement a time out for the challenge.

### CODE REVIEW

File: contracts/src/Rollup.sol [hunk 1]
```diff
@@ -314,22 +314,21 @@ contract Rollup is RollupBase {
...
+        }
+        if (challengerAssertionID > lastCreatedAssertionID) {
+            revert UnproposedAssertion();
+        }
+        if (lastConfirmedAssertionID >= defenderAssertionID) {
+            revert AssertionAlreadyResolved();
+        }
+        // Require that players have attested to sibling assertions.
+        if (parentID != assertions[challengerAssertionID].parent) {
+            revert NotSiblings();
```
Comment (ujvl) [inline contracts/src/Rollup.sol:331]: the scope definition here was made to get around another EVM quirk: there are limitations on the structure of the stack https://soliditydeveloper.com/stacktoodeep

this change is fine as long as it compiles

File: contracts/src/challenge/SymChallenge.sol [hunk 2]
```diff
@@ -86,23 +86,46 @@ contract SymChallenge is ChallengeBase, ISymChallenge {
...
         if (bisectionHash != 0) {
             revert AlreadyInitialized();
         }
         require(_numSteps > 0, "INVALID_NUM_STEPS");
-        bisectionHash = ChallengeLib.initialBisectionHash(startStateHash, endStateHash, _numSteps);
-        // TODO: consider emitting a different event?
-        emit Bisected(bisectionHash, 0, _numSteps);
+
+        // Defender proposes `numSteps` and `endStateHash` first
+        if (turn == Turn.Defender && msg.sender == defender) {
```
Comment (ujvl) [inline contracts/src/challenge/SymChallenge.sol:104]: should we use the `onlyOnTurn` modifier in the function definition? then you can just check
`if (turn == Turn.Defender)` and also don't need to do the reverts below

File: contracts/src/challenge/SymChallenge.sol [hunk 3]
```diff
@@ -86,23 +86,46 @@ contract SymChallenge is ChallengeBase, ISymChallenge {
...
 
         turn = Turn.Defender;
         lastMoveBlock = block.number;
         defenderTimeLeft = challengePeriod;
         challengerTimeLeft = challengePeriod;
     }
 
-    // TODO: use minimum of challenger/defender's proposed numSteps.
-    function initializeChallengeLength(uint256 _numSteps) external override onlyOnTurn {
+    function proposeChallenge(bytes32 _endStateHash, uint256 _numSteps) external {
```
Comment (ujvl) [inline contracts/src/challenge/SymChallenge.sol:96]: what happens in the special case where they both pass in the same hashes but for different `numSteps`:
```
// defender calls:
proposeChallenge(sameStateHash, 32)
// challenger calls:
proposeChallenge(sameStateHash, 64)
```

File: contracts/src/challenge/SymChallenge.sol [hunk 4]
```diff
@@ -55,26 +56,25 @@ contract SymChallenge is ChallengeBase, ISymChallenge {
...
      */
+
     function initialize(
         address _defender,
         address _challenger,
         IVerifier _verifier,
         IDAProvider _daProvider,
         IChallengeResultReceiver _resultReceiver,
         bytes32 _startStateHash,
-        bytes32 _endStateHash,
```
Comment (ujvl) [inline contracts/src/challenge/SymChallenge.sol:74]: one issue here is a malicious defender can propose the correct hash—which isn't consistent with what the assertion committed to, but they nevertheless are able to win the challenge

File: contracts/src/challenge/SymChallenge.sol [hunk 5]
```diff
@@ -149,6 +191,8 @@ contract SymChallenge is ChallengeBase, ISymChallenge {
         emit Bisected(bisectionHash, challengedSegmentStart, challengedSegmentLength);
     }
 
+    // TODO: If bisection is not executed because both have commited to the same has at the same `numStep`, verify
+    //        there are no additional ops on the stack after the last step.
```
Comment (ujvl) [inline contracts/src/challenge/SymChallenge.sol:195]: typo in `"same has"` (i assume you mean hash) + indentation looks off

**Review by ujvl — APPROVED** [review]

File: contracts/src/challenge/SymChallenge.sol [hunk 6]
```diff
@@ -55,26 +58,30 @@ contract SymChallenge is ChallengeBase, ISymChallenge {
...
      * @param _verifier Address of the verifier contract.
      * @param _daProvider DA provider.
      * @param _resultReceiver Address of contract that will receive the outcome (via callback `completeChallenge`).
      * @param _startStateHash Commitment to agreed-upon start state.
-     * @param _endStateHash Commitment to disagreed-upon end state being challenged.
+     * @param _endStateDefenseHash Proposed DA end state hash of the defender.
+     * @param _endStateChallengeHash Proposed DA end state hash of the challenger.
+     * @param challengePeriod The duration that challenger/defender has for continuing the challenge
      */
+
```
Comment (ujvl) [inline contracts/src/challenge/SymChallenge.sol:72]: nit: extraneous newline between documentation and fn definition

Comment (chalkandpaste) [reply contracts/src/challenge/SymChallenge.sol]: Fixed

File: contracts/src/challenge/SymChallenge.sol (hunk 5 above)
Comment (chalkandpaste) [reply contracts/src/challenge/SymChallenge.sol]: Fixed.

**Pull request merged**